### PR TITLE
Vulns, in JSON!

### PR DESCRIPTION
**Pull request by DarthHater**
This is a start of adding the ability to serialize and deserialize vulnerabilities in the CycloneDX bom, in JSON!

### CODE REVIEW

File: src/test/java/org/cyclonedx/BomJsonGeneratorTest.java [hunk 1]
```diff
@@ -46,6 +46,17 @@ public void schema12GenerationTest() throws Exception {
...
 
+    @Test
+    public void schema12GenerationTestWithVulns() throws Exception {
+        Bom bom = createCommonBom("/bom-1.2-vulnerability.xml");
+        BomJsonGenerator generator = BomGeneratorFactory.createJson(Version.VERSION_12, bom);
+        Assert.assertTrue(generator instanceof BomJsonGenerator12);
+        Assert.assertEquals(CycloneDxSchema.Version.VERSION_12, generator.getSchemaVersion());
+        File file = writeToFile(generator.toJsonString());
+        JsonParser parser = new JsonParser();
+        Assert.assertTrue(parser.isValid(file, CycloneDxSchema.Version.VERSION_12, true));
```
Comment (DarthHater) [inline src/test/java/org/cyclonedx/BomJsonGeneratorTest.java:57]: Failing atm, we need to register a vulns schema or something akin so it knows it's good to go

File: src/test/java/org/cyclonedx/parsers/JsonParserTest.java [hunk 2]
```diff
@@ -155,6 +155,14 @@ public void testParsedObjects12Bom() throws Exception {
         Assert.assertNull(d11.getDependencies());
     }
 
+    @Test
+    public void testValid12BomWithVulnerabilities() throws Exception {
+        final byte[] bomBytes = IOUtils.toByteArray(this.getClass().getResourceAsStream("/bom-1.2-vulnerability.json"));
+        final JsonParser parser = new JsonParser();
+        final Bom bom = parser.parse(bomBytes);
+        Assert.assertNotNull(bom);
```
Comment (DarthHater) [inline src/test/java/org/cyclonedx/parsers/JsonParserTest.java:163]: This is mostly a stub test to make sure things occur correctly, it needs to be filled out in much more detail

File: src/test/resources/bom-1.2-vulnerability.xml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (DarthHater) [inline src/test/resources/bom-1.2-vulnerability.xml:1]: This is not a 1.2 schema, just a copy of 1.1 with a vuln, so it'd make sense to actually make it a 1.2 schema and bring in more bits, merge it with having a vuln, etc...

File: src/main/java/org/cyclonedx/generators/json/AbstractBomJsonGenerator.java [hunk 4]
```diff
@@ -54,6 +56,9 @@ private void setupObjectMapper(final ObjectMapper mapper) {
         SimpleModule depModule = new SimpleModule();
         SimpleModule componentWrapperModule = new SimpleModule();
 
+        mapper.setAnnotationIntrospector(
+            new XmlOnlyAnnotationIntrospector());
```
Comment (DarthHater) [inline src/main/java/org/cyclonedx/generators/json/AbstractBomJsonGenerator.java:60]: New annotation introspector to exclude fields when generating JSON, that are stuff like the namespace, etc... that show up in XML but not JSON

**Comment by stevespringett** [comment]
I won't be able to merge this as the schema for JSON vulnerabilities has not been defined yet. It will also fail validation when using the JSON strict schema.

See https://github.com/CycloneDX/specification/issues/37

Prior to adding support to official implementations, an approach to supporting extensions for JSON needs to be agreed upon and the schema has been adopted through the [standardization process](https://cyclonedx.org/about/standardization-process/).

Additionally, there are discussions in the community for advancing the vulnerability extension.
See: 
* https://github.com/CycloneDX/specification/pull/44
* https://github.com/CycloneDX/specification/issues/38

**Comment by DarthHater** [comment]
Sounds good @stevespringett , @bhamail and I had generated a JSON schema way back when we first looked at this, I'll see if I can resurrect it and add it to the discussion

**Comment by DarthHater** [comment]
Closing this, it's old!

**Pull request closed without merging**